castget.1
.TH "CASTGET" "1" "January 2025" "castget 2.0.1" "User Commands"
.
.SH "NAME"
\fBcastget\fR \- download RSS enclosures
.
.SH "SYNOPSIS"
\fBcastget\fR [\fIOPTION\fR\.\.\.] [\fICHANNEL IDENTIFIER\fR\.\.\.]
.
.SH "DESCRIPTION"
\fBcastget\fR downloads RSS enclosures, for example podcasts, from RSS feeds\.
.
.P
RSS feeds are assigned \fIchannel identifier\fRs in the configuration file \fB\.castgetrc\fR\. Channels to be processed by \fBcastget\fR are indicated by supplying the channel identifiers as arguments\. If no channel identifiers are provided, all available channels are processed\.
.
.P
If run without any options, \fBcastget\fR will perform the default action on all channels to be processed\. The default action is to download any enclosure not already downloaded\. Other actions may be performed by supplying one or more options as arguments\.
.
.SH "OPTIONS"
.
.SS "Operations"
.
.TP
\fB\-c\fR, \fB\-\-catchup\fR
Catch up with channels and exit\.
.
.TP
\fB\-l\fR, \fB\-\-list\fR
List available enclosures that have not yet been downloaded, and exit\.
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help and exit\.
.
.TP
\fB\-V\fR, \fB\-\-version\fR
Output version information and exit\.
.
.SS "Operation filters"
.
.TP
\fB\-n\fR, \fB\-\-new\-only\fR
Restrict operation to new channels only, i\.e\. to channels that have never been downloaded from or been caught up with before\. Note that if a channel is added to the configuration and subsequently removed, its download history is preserved\. This means that a channel that has been removed from the configuration file will not be considered as \'new\' if it is added to the configuration again at a later time\.
.
.TP
\fB\-1\fR, \fB\-\-first\-only\fR
Restrict operation to the most recent item in each channel only\.
.
.TP
\fB\-f\fR \fIpattern\fR, \fB\-\-filter\fR=\fIpattern\fR
Restrict operation to enclosures whose download URLs match the regular expression \fBpattern\fR\. Note that this will override any regular expression filters given in the configuration file\.
.
.SS "Global options"
.
.TP
\fB\-r\fR, \fB\-\-resume\fR
Resume aborted downloads\. Make sure not to use this option if the RSS feed uses the same filename for multiple enclosures as this will corrupt existing downloads\.
.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print anything except error messages\.
.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print detailed progress information\.
.
.TP
\fB\-d\fR, \fB\-\-debug\fR
Print (lots of) connection debug information\.
.
.TP
\fB\-p\fR, \fB\-\-progress\-bar\fR
Print a progress bar when downloading enclosures\.
.
.TP
\fB\-C\fR \fIfilename\fR, \fB\-\-rcfile\fR=\fIfilename\fR
Override the default filename for the configuration file\.
.
.SH "EXAMPLES"
.
.TP
Download all enclosures not already downloaded:
.
.IP
$ castget
.
.TP
Download all enclosures not already downloaded from channel \fBfoobar\fR and be verbose:
.
.IP
$ castget \-v foobar
.
.TP
List all enclosures not already downloaded:
.
.IP
$ castget \-l
.
.TP
Catch up with channel \fBfoobar\fR:
.
.IP
$ castget \-c foobar
.
.TP
List the first available enclosure in all new channels:
.
.IP
$ castget \-1 \-n \-l
.
.TP
Catch up with items that match the regular expression \fBFreddies0[67]\fR in the channel \fBfrederator\fR:
.
.IP
$ castget \-c \-f "Freddies0[67]" frederator
.
.SH "HTTP PROXY"
.
.TP
To use a HTTP proxy, set the environment variable \fBhttp_proxy\fR:
.
.IP
http_proxy=http://your\.proxy\.server:port/ castget
.
.SH "SEE ALSO"
castgetrc(5)
.
.SH "BUGS"
Please see the castget home page \fIhttp://mlj\.github\.io/castget\fR for instructions on how to submit bug reports\.
.
.SH "AUTHORS"
Marius L\. Jøhndal, Jick Nan\.
.
.SH "COPYRIGHT"
Castget is Copyright (C) 2005\-2021 Marius L\. Jøhndal\.
.
.P
Castget is Copyright (C) 2007 Jick Nan\.
.
.P
This is free software; see the source for copying conditions\. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\.


NEWS
Version 1.1.0:

  * New command line option --rcfile for specifying an alternate location for
  the configuration file.

  * New command line option --filter and configuration key filter for
  restricting operations to enclosures matching a regular
  expression. This feature requires glib >= 2.14 and can be disabled
  by supplying --disable-gregex to configure.

  * Distribution files are compressed with bzip2.

Version 1.0.1:

  * castget now identifies itself using the HTTP User-Agent header.

Version 1.0.0:

  * First stable release.


src/castget.c
/*
  Copyright (C) 2005-2021 Marius L. Jøhndal

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include "channel.h"
#include "configuration.h"

#define _GNU_SOURCE
#include <errno.h>
#include <getopt.h>
#include <glib/gprintf.h>
#include <glib/gstdio.h>
#include <libxml/parser.h>
#include <string.h>
#include <unistd.h>
#ifdef HAVE_TAGLIB
#include <taglib/tag_c.h>
#endif /* HAVE_TAGLIB */

enum op { OP_UPDATE, OP_CATCHUP, OP_LIST };

static int _process_channel(const gchar *channel_directory, GKeyFile *kf,
                            const char *identifier, enum op op,
                            struct channel_configuration *defaults,
                            enclosure_filter *filter);
static void version(void);
static GKeyFile *_configuration_file_open(const gchar *rcfile);
static void _configuration_file_close(GKeyFile *kf);
#ifdef HAVE_TAGLIB
static void _set_tags(const gchar *filename,
                      const struct channel_configuration *cfg);
#endif /* HAVE_TAGLIB */
static int playlist_add(const gchar *playlist_file, const gchar *media_file);

static gboolean verbose = FALSE;
static gboolean quiet = FALSE;
static gboolean new_only = FALSE;
static option_info *opts = NULL;

int main(int argc, char **argv)
{
  enum op op = OP_UPDATE;
  int i;
  int ret = 0;
  gchar **groups;
  gchar *channeldir;
  GKeyFile *kf;
  struct channel_configuration *defaults;
  enclosure_filter *filter = NULL;
  GError *error = NULL;
  GOptionContext *context;
  static gboolean first_only = FALSE;
  static gboolean first_only_disregard_eligibility = FALSE;
  static gint stop_after_count = 0;
  static gboolean count_disregards_eligibility = FALSE;
  static gboolean resume = FALSE;
  static gboolean debug = FALSE;
  static gboolean reverse = FALSE;
  static gboolean show_progress_bar = FALSE;
  static gchar *rcfile = NULL;
  static gchar *filter_regex = NULL;
  static gboolean list = FALSE;
  static gboolean catchup = FALSE;
  static gboolean show_version = FALSE;

  static GOptionEntry options[] = {
    { "catchup", 'c', 0, G_OPTION_ARG_NONE, &catchup,
      "catch up with channels and exit" },
    { "list", 'l', 0, G_OPTION_ARG_NONE, &list,
      "list available enclosures that have not yet been downloaded and exit" },
    { "version", 'V', 0, G_OPTION_ARG_NONE, &show_version,
      "print version and exit" },

    { "resume", 'r', 0, G_OPTION_ARG_NONE, &resume,
      "resume aborted downloads" },
    { "rcfile", 'C', 0, G_OPTION_ARG_FILENAME, &rcfile,
      "override the default configuration file name" },

    { "debug", 'd', 0, G_OPTION_ARG_NONE, &debug,
      "print connection debug information" },
    { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose,
      "print detailed progress information" },
    { "progress-bar", 'p', 0, G_OPTION_ARG_NONE, &show_progress_bar,
      "print progress bar" },

    { "new-only", 'n', 0, G_OPTION_ARG_NONE, &new_only,
      "only process new channels" },
    { "quiet", 'q', 0, G_OPTION_ARG_NONE, &quiet, "only print error messages" },
    { "stop-after", 's', 0, G_OPTION_ARG_INT, &stop_after_count,
      "stop after processing ARGUMENT items from each channel" },
    { "count-disregards-eligibility", 'x', 0, G_OPTION_ARG_NONE, &count_disregards_eligibility,
      "when processing --stop-after, count considers all items, not just those eligible" },
    { "first-only", '1', 0, G_OPTION_ARG_NONE, &first_only,
      "only process the most recent item from each channel" },
    { "first-only-disregard-eligibility", 'c', 0, G_OPTION_ARG_NONE, &first_only_disregard_eligibility,
      "only process the most recent item from each channel" },
    { "reverse", 'R', 0, G_OPTION_ARG_NONE, &reverse,
      "process the channel in reverse order" },
    { "filter", 'f', 0, G_OPTION_ARG_STRING, &filter_regex,
      "only process items whose enclosure names match a regular expression" },

    { NULL }
  };

  context = g_option_context_new("CHANNELS");
  g_option_context_add_main_entries(context, options, NULL);

  if (!g_option_context_parse(context, &argc, &argv, &error)) {
    g_print("option parsing failed: %s\n", error->message);
    exit(1);
  }

  /* Do some additional sanity checking of options. */
  if ((verbose && quiet) || (show_progress_bar && quiet)) {
    g_print("option parsing failed: options are incompatible.\n");
    exit(1);
  }

  if ((catchup && list) || (catchup && show_version) ||
      (list && show_version)) {
    g_print(
        "option parsing failed: --catchup, --list and --version options are "
        "incompatible.\n");
    exit(1);
  }

  if (stop_after_count < 0) {
    g_print("--stop-after must be greater than 0\n");
    exit(1);
  }
  if (first_only && stop_after_count) {
    g_print("--stop-after and --first-only are incompatible\n");
    exit(1);
  }
  if (first_only_disregard_eligibility && stop_after_count) {
    g_print("--stop-after and --first-only-disregard-eligibility are incompatible\n");
    exit(1);
  }
  if (first_only_disregard_eligibility && first_only) {
    g_print("--first-only and --first-only-disregard-eligibility are incompatible\n");
    exit(1);
  }

  if (first_only)
    stop_after_count = 1;
  if (first_only_disregard_eligibility) {
    stop_after_count = 1;
    count_disregards_eligibility = TRUE;
  }

  /* Decide on the action to take */
  if (show_version) {
    version();
    exit(0);
  }

  opts = option_info_new();
  opts->no_download = 0;
  opts->no_mark_read = 0;
  opts->stop_after_count = stop_after_count;
  opts->count_disregards_eligibility = count_disregards_eligibility;
  opts->resume = resume;
  opts->debug = debug;
  opts->reverse = reverse;
  opts->show_progress_bar = show_progress_bar;

  if (catchup) {
    op = OP_CATCHUP;
    opts->no_download = 1;
    opts->no_mark_read = 0;
  }

  if (list) {
    op = OP_LIST;
    opts->no_download = 1;
    opts->no_mark_read = 1;
  }

  opts->filter = NULL;
  if (filter_regex) {
    filter = enclosure_filter_new(filter_regex, FALSE);
    opts->filter = filter;
    g_free(filter_regex);
  }

  if (verbose && new_only)
    g_print("Fetching new channels only...\n");

  LIBXML_TEST_VERSION;

  /* Build the channel directory path and ensure that it exists. */
  channeldir = g_build_filename(g_get_home_dir(), ".castget", NULL);

  if (!g_file_test(channeldir, G_FILE_TEST_IS_DIR)) {
    if (g_mkdir(channeldir, 0755) < 0) {
      perror("Error creating channel directory");
      return 1;
    }
  }

  /* Try opening configuration file. */
  if (!rcfile)
    /* Supply default path name. */
    rcfile = g_build_filename(g_get_home_dir(), ".castgetrc", NULL);

  kf = _configuration_file_open(rcfile);

  if (kf) {
    /* Read defaults. */
    if (g_key_file_has_group(kf, "*")) {
      /* Verify the keys in the global configuration. */
      if (channel_configuration_verify_keys(kf, "*") < 0)
        return -1;

      defaults = channel_configuration_new(kf, "*", NULL);
    } else
      defaults = NULL;

    /* Perform actions. */
    if (optind < argc) {
      while (optind < argc)
        _process_channel(channeldir, kf, argv[optind++], op, defaults, filter);
    } else {
      groups = g_key_file_get_groups(kf, NULL);

      for (i = 0; groups[i]; i++)
        if (strcmp(groups[i], "*"))
          _process_channel(channeldir, kf, groups[i], op, defaults, filter);

      g_strfreev(groups);
    }

    /* Clean up defaults. */
    if (defaults)
      channel_configuration_free(defaults);
  } else
    ret = 1;

  /* Clean-up. */
  g_free(channeldir);

  if (filter)
    enclosure_filter_free(filter);

  option_info_free(opts);

  g_free(rcfile);

  if (kf)
    _configuration_file_close(kf);

  xmlCleanupParser();

  return ret;
}

static void version(void)
{
  g_printf("%s %s", PACKAGE, VERSION);
#ifdef HAVE_TAGLIB
  g_printf(" with taglib support\n");
#else
  g_printf("\n");
#endif

  g_printf("Copyright (C) 2005-2021 Marius L. Jøhndal\n");
}

static void _print_item_update(const enclosure *enclosure,
                               const gchar *filename)
{
  if (enclosure->length > 0) {
    gchar *size = g_format_size(enclosure->length);
    g_printf(" * %s (%s)\n", enclosure->url, size);
    g_free(size);
  } else
    g_printf(" * %s (unknown size)\n", filename);
}

static void update_callback(void *user_data, channel_action action,
                            channel_info *channel_info, enclosure *enclosure,
                            const gchar *filename)
{
  struct channel_configuration *c = (struct channel_configuration *)user_data;

  switch (action) {
  case CCA_RSS_DOWNLOAD_START:
    if (!quiet)
      g_printf("Updating channel %s...\n", c->identifier);
    break;

  case CCA_RSS_DOWNLOAD_END:
    break;

  case CCA_ENCLOSURE_DOWNLOAD_START:
    g_assert(channel_info);
    g_assert(enclosure);

    if (verbose)
      _print_item_update(enclosure, filename);

    break;

  case CCA_ENCLOSURE_DOWNLOAD_END:
    g_assert(channel_info);
    g_assert(enclosure);
    g_assert(filename);

#ifdef HAVE_TAGLIB
    _set_tags(filename, c);
#endif /* HAVE_TAGLIB */

    /* Update playlist. */
    if (c->playlist) {
      playlist_add(c->playlist, filename);

      if (verbose)
        printf(" * Added downloaded enclosure %s to playlist %s.\n", filename,
               c->playlist);
    }
    break;
  }
}

static void catchup_callback(void *user_data, channel_action action,
                             channel_info *channel_info, enclosure *enclosure,
                             const gchar *filename)
{
  struct channel_configuration *c = (struct channel_configuration *)user_data;

  switch (action) {
  case CCA_RSS_DOWNLOAD_START:
    if (!quiet)
      g_printf("Catching up with channel %s...\n", c->identifier);
    break;

  case CCA_RSS_DOWNLOAD_END:
    break;

  case CCA_ENCLOSURE_DOWNLOAD_START:
    g_assert(channel_info);
    g_assert(enclosure);

    if (verbose)
      _print_item_update(enclosure, filename);

    break;

  case CCA_ENCLOSURE_DOWNLOAD_END:
    break;
  }
}

static void list_callback(void *user_data, channel_action action,
                          channel_info *channel_info, enclosure *enclosure,
                          const gchar *filename)
{
  struct channel_configuration *c = (struct channel_configuration *)user_data;

  switch (action) {
  case CCA_RSS_DOWNLOAD_START:
    g_printf("Listing channel %s...\n", c->identifier);
    break;

  case CCA_RSS_DOWNLOAD_END:
    break;

  case CCA_ENCLOSURE_DOWNLOAD_START:
    g_assert(channel_info);
    g_assert(enclosure);

    if (verbose)
      _print_item_update(enclosure, filename);

    break;

  case CCA_ENCLOSURE_DOWNLOAD_END:
    break;
  }
}

static int _process_channel(const gchar *channel_directory, GKeyFile *kf,
                            const char *identifier, enum op op,
                            struct channel_configuration *defaults,
                            enclosure_filter *filter)
{
  channel *c;
  gchar *channel_filename, *channel_file;
  struct channel_configuration *channel_configuration;
  enclosure_filter *per_channel_filter = NULL;

  /* Check channel identifier and read channel configuration. */
  if (!g_key_file_has_group(kf, identifier)) {
    fprintf(stderr, "Unknown channel identifier %s.\n", identifier);

    return -1;
  }

  /* Verify the keys in the channel configuration. */
  if (channel_configuration_verify_keys(kf, identifier) < 0)
    return -1;

  channel_configuration = channel_configuration_new(kf, identifier, defaults);

  /* Check that mandatory keys were set. */
  if (!channel_configuration->url) {
    fprintf(stderr, "No feed URL set for channel %s.\n", identifier);

    channel_configuration_free(channel_configuration);
    return -1;
  }

  if (!channel_configuration->spool_directory) {
    fprintf(stderr, "No spool directory set for channel %s.\n", identifier);

    channel_configuration_free(channel_configuration);
    return -1;
  }

  /* Construct channel file name. */
  channel_filename = g_strjoin(".", identifier, "xml", NULL);
  channel_file = g_build_filename(channel_directory, channel_filename, NULL);
  g_free(channel_filename);

  if (new_only && access(channel_file, F_OK) == 0) {
    /* If we are only fetching new channels, skip the channel if there is
       already a channel file present. */

    channel_configuration_free(channel_configuration);
    return 0;
  }

  c = channel_new(channel_configuration->url, channel_file,
                  channel_configuration->spool_directory,
                  channel_configuration->filename_pattern, opts->resume);
  g_free(channel_file);

  if (!c) {
    fprintf(stderr, "Error parsing channel file for channel %s.\n", identifier);

    channel_configuration_free(channel_configuration);
    return -1;
  }

  /* Set up per-channel filter unless overridden on the command
     line. */
  if (!filter && channel_configuration->regex_filter) {
    per_channel_filter =
        enclosure_filter_new(channel_configuration->regex_filter, FALSE);

    filter = per_channel_filter;
  }

  switch (op) {
  case OP_UPDATE:
    channel_update(c, channel_configuration, update_callback, opts);
    break;

  case OP_CATCHUP:
    channel_update(c, channel_configuration, catchup_callback, opts);
    break;

  case OP_LIST:
    channel_update(c, channel_configuration, list_callback, opts);
    break;
  }

  /* Clean-up. */
  if (per_channel_filter)
    enclosure_filter_free(per_channel_filter);

  channel_free(c);
  channel_configuration_free(channel_configuration);

  return 0;
}

static GKeyFile *_configuration_file_open(const gchar *rcfile)
{
  GKeyFile *kf;
  GError *error = NULL;

  kf = g_key_file_new();

  if (!g_key_file_load_from_file(kf, rcfile, G_KEY_FILE_NONE, &error)) {
    fprintf(stderr, "Error reading configuration file %s: %s.\n", rcfile,
            error->message);
    g_error_free(error);
    g_key_file_free(kf);
    kf = NULL;
  }

  return kf;
}

static void _configuration_file_close(GKeyFile *kf)
{
  g_key_file_free(kf);
}

#ifdef HAVE_TAGLIB
static void _set_tags(const gchar *filename,
                      const struct channel_configuration *cfg)
{
  if (cfg->artist_tag || cfg->title_tag || cfg->album_tag || cfg->genre_tag ||
      cfg->year_tag || cfg->comment_tag) {
    TagLib_File *file;
    TagLib_Tag *tag;

    file = taglib_file_new(filename);

    if (file == NULL) {
      fprintf(stderr, "Error setting tags for file %s.\n", filename);
      return;
    }

    tag = taglib_file_tag(file);

    if (cfg->artist_tag) {
      taglib_tag_set_artist(tag, cfg->artist_tag);

      if (verbose)
        printf(" * Set artist tag to %s.\n", cfg->artist_tag);
    }

    if (cfg->title_tag) {
      taglib_tag_set_title(tag, cfg->title_tag);

      if (verbose)
        printf(" * Set title tag to %s.\n", cfg->title_tag);
    }

    if (cfg->album_tag) {
      taglib_tag_set_album(tag, cfg->album_tag);

      if (verbose)
        printf(" * Set album tag to %s.\n", cfg->album_tag);
    }

    if (cfg->genre_tag) {
      taglib_tag_set_genre(tag, cfg->genre_tag);

      if (verbose)
        printf(" * Set genre tag to %s.\n", cfg->genre_tag);
    }

    if (cfg->year_tag) {
      taglib_tag_set_year(tag, g_ascii_strtoull(cfg->year_tag, NULL, 10));

      if (verbose)
        printf(" * Set year tag to %s.\n", cfg->year_tag);
    }

    if (cfg->comment_tag) {
      taglib_tag_set_comment(tag, cfg->comment_tag);

      if (verbose)
        printf(" * Set comment tag to %s.\n", cfg->comment_tag);
    }

    if (!taglib_file_save(file))
      fprintf(stderr, "Error setting tags for file %s.\n", filename);

    taglib_tag_free_strings();
    taglib_file_free(file);
  }
}
#endif /* HAVE_TAGLIB */

static int playlist_add(const gchar *playlist_file, const gchar *media_file)
{
  FILE *f;

  f = fopen(playlist_file, "a");

  if (!f) {
    fprintf(stderr, "Error opening playlist file %s: %s.\n", playlist_file,
            strerror(errno));
    return -1;
  }

  fprintf(f, "%s\n", media_file);
  fclose(f);
  return 0;
}
